src/table/memo.rs

```diff
@@ -22,6 +24,71 @@ pub(crate) trait Memo: Any + Send + Sync + Debug {
     fn origin(&self) -> &QueryOrigin;
 }
 
+/// An untyped memo that can only be dropped.
+pub struct MemoDrop(ManuallyDrop<Arc<DummyMemo>>, unsafe fn(Arc<DummyMemo>));
+
+impl MemoDrop {
+    pub fn new<M: Memo>(memo: Arc<M>) -> Self {
+        Self(
+            ManuallyDrop::new(to_dummy::<M>(memo)),
+            // SAFETY: `M` is the same as used in `to_dummy`
+            |memo| unsafe {
+                from_dummy::<M>(memo);
+            },
+        )
+    }
+}
+
+impl Drop for MemoDrop {
+    fn drop(&mut self) {
+        // SAFETY: We only construct this type with a valid drop function pointer
+        unsafe { self.1(ManuallyDrop::take(&mut self.0)) };
+    }
+}
+
+pub fn spawn_memo_drop_thread(receiver: MemoDropReceiver) -> std::thread::JoinHandle<()> {
+    std::thread::spawn(|| {
+        receiver.0.into_iter().for_each(|e| match e {
+            MemoDropAction::Drop(memo_drop) => drop(memo_drop),
+            // We were instructed to park the thread, this means we have dropped everything for the
+            // last revision.
+            MemoDropAction::Park(parker) => parker.park(),
+        });
+    })
+}
+
+pub fn memo_drop_channel() -> (MemoDropSender, MemoDropReceiver) {
+    let (tx, rx) = std::sync::mpsc::channel();
```

Comment:
Should we use a `sync_channel` instead to apply back pressure if the drop thread is getting overwhelmed?

What happens if a drop implementatio panics? I assume that leaves the entire database in a poisoned state because the  drop thread is now "broken"?

Reply:
Good question regarding backpressure, unsure :)

> What happens if a drop implementatio panics? I assume that leaves the entire database in a poisoned state because the drop thread is now "broken"?

Bad things. I have actually forgotten about destructors panicking, right now that will tear down the dropper thread and cause us to panic when evicting. It should be simple enough to workaround though with a `catch_unwind`, the bigger question then though becomes if its fine for us to silently eat panic payloads in destructors ...